File: src/main/java/cad97/spawnercraft/SpawnerCraft.java
package cad97.spawnercraft;

import cad97.spawnercraft.proxy.IProxy;
import cad97.spawnercraft.utility.LogHelper;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;

@SuppressWarnings("DefaultAnnotationParam")
@Mod(modid = SpawnerCraft.MOD_ID, name = SpawnerCraft.MOD_NAME, version = SpawnerCraft.VERSION,
        guiFactory = SpawnerCraft.GUI_FACTORY_CLASS, dependencies = SpawnerCraft.DEPENDENCIES,
        acceptedMinecraftVersions = SpawnerCraft.ACCEPTED_MINECRAFT_VERSION)
public class SpawnerCraft {

    public static final String MOD_ID = "spawnercraft";
    public static final String MOD_NAME = "SpawnerCraft";

    static final String VERSION = "4.0.0";
    static final String GUI_FACTORY_CLASS = "cad97.spawnercraft.client.gui.GuiFactory"; // GuiFactory.class.getCanonicalName()
    static final String DEPENDENCIES = "";
    static final String ACCEPTED_MINECRAFT_VERSION = "[1.12]";

    @Mod.Instance
    static SpawnerCraft instance;

    @SidedProxy(clientSide = "cad97.spawnercraft.proxy.ClientProxy", serverSide = "cad97.spawnercraft.proxy.ServerProxy")
    private static IProxy proxy;

    @Mod.EventHandler
    public void preInit(FMLPreInitializationEvent event) {
        LogHelper.setLogger(event.getModLog());
        proxy.preInit(event);
        LogHelper.logInfo("preInit finished.");
    }

    @Mod.EventHandler
    public void init(FMLInitializationEvent event) {
        proxy.init(event);
        LogHelper.logInfo("init finished.");
    }

    @Mod.EventHandler
    public void postInit(FMLPostInitializationEvent event) {
        proxy.postInit(event);
        LogHelper.logInfo("postInit finished.");
    }
}

File: src/main/kotlin/cad97/spawnercraftkt/SpawnerCraft.kt
package cad97.spawnercraftkt

import cad97.spawnercraftkt.gui.GuiFactory
import cad97.spawnercraftkt.proxy.ClientProxy
import cad97.spawnercraftkt.proxy.Proxy
import cad97.spawnercraftkt.proxy.ServerProxy
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.SidedProxy
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import org.apache.logging.log4j.Logger

@Mod(
        modid = SpawnerCraft.modid,
        version = SpawnerCraft.version,
        name = SpawnerCraft.name,
        guiFactory = GuiFactory.qualifiedName,
        useMetadata = SpawnerCraft.useMetadata,
        acceptedMinecraftVersions = SpawnerCraft.acceptedMinecraftVersions
)
class SpawnerCraft {
    companion object {
        @Mod.Instance
        lateinit var instance: SpawnerCraft private set
        @SidedProxy(modId = modid, clientSide = ClientProxy.qualifiedName, serverSide = ServerProxy.qualifiedName)
        lateinit var proxy: Proxy private set
        lateinit var logger: Logger private set
        const val modid = "spawnercraftkt"
        const val name = "SpawnerCraft"
        const val version = "<%version%>"
        const val useMetadata = true
        const val acceptedMinecraftVersions = "1.12"
    }

    @Mod.EventHandler
    fun preInit(event: FMLPreInitializationEvent) {
        logger = event.modLog
        proxy.preInit(event)
        logger.info("PreInitialization complete")
    }

    @Mod.EventHandler
    fun init(event: FMLInitializationEvent) {
        proxy.init(event)
        logger.info("Initialization complete")
    }

    @Mod.EventHandler
    fun postInit(event: FMLPostInitializationEvent) {
        proxy.postInit(event)
        logger.info("PostInitialization complete")
    }
}
